### PR TITLE
Residualized Intermediate Representation (RIR)

### PR DESCRIPTION
**Pull request by cesarzc**

### CODE REVIEW

File: compiler/qsc_rir/src/rir.rs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    /// The input type of the callable.
+    pub input_type: Vec<Ty>,
+    /// The output type of the callable.
+    pub output_type: Option<Ty>,
+    /// The callable body.
+    /// N.B. `None` bodys represent an intrinsic.
+    pub body: Option<BlockId>,
+}
+
+pub enum Instruction {
```
Comment (swernli) [inline compiler/qsc_rir/src/rir.rs:39]: Thinking about it some more, I think removing the concept of statement means some instructions need to have a Variable field for their result. Take LogicalNot as an example. Where does the result of that instruction get written to? I think Call will need an Option<Variable> and others that always produce a value will need Variable as a field.

**Review by swernli — COMMENTED** [review]
Just needs a couple tweaks!

File: compiler/qsc_rir/src/rir.rs (hunk 1 above)
Comment (cesarzc) [reply compiler/qsc_rir/src/rir.rs:39]: Good catch! I'll update the instructions that need a variable field.

File: compiler/qsc_rir/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,7 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#![warn(clippy::mod_module_files, clippy::pedantic, clippy::unwrap_used)]
```
[the remark this answers is not in the document]
Comment (cesarzc) [reply compiler/qsc_rir/src/lib.rs]: Updated.

File: compiler/qsc_rir/src/rir.rs (hunk 1 above)
Comment (cesarzc) [reply compiler/qsc_rir/src/rir.rs:39]: Updated.

**Review by swernli — APPROVED** [review]

**Pull request merged**